projects/chaos/src/relay_fsm.c

```diff
@@ -0,0 +1,154 @@
+#include "relay_fsm.h"
+
+#include <stdbool.h>
+#include <stdint.h>
+#include <string.h>
+
+#include "can.h"
+#include "can_ack.h"
+#include "can_msg_defs.h"
+#include "can_transmit.h"
+#include "chaos_events.h"
+#include "event_queue.h"
+#include "fsm.h"
+#include "misc.h"
+#include "relay_id.h"
+#include "status.h"
+
+#define RELAY_FSM_MAX_RETRIES 3
+
+typedef struct RelayFsmAckCtx {
+  RelayId id;
+  uint8_t retries;
+} RelayFsmAckCtx;
+
+typedef struct RelayFsmCtx {
+  RelayFsmAckCtx ack_ctx;
+  CANAckRequest request;
```

Comment:
Ah, you're just pre-allocating the ack request here so you don't have to create one in `prv_relay_transmit`?

Reply:
Exactly.